crypto/merklekeystore/keystore_test.go

```diff
@@ -0,0 +1,538 @@
+	hashable, sig := makeSig(signer, start, a)
+
+	sig2 := sig
+	bs := make([]byte, len(sig.ByteSignature))
+	copy(bs, sig2.ByteSignature)
+	bs[0]++
+	sig2.ByteSignature = bs
+	a.Error(signer.GetVerifier().Verify(start, start, 1, hashable, sig2))
+}
+
+func TestAttemptToUseDifferentKey(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	a := require.New(t)
+	var err error
+
+	start, _, signer := generateTestSignerAux(a)
+	defer signer.keyStore.store.Close()
+	hashable, sig := makeSig(signer, start+1, a)
+	// taking signature for specific round and changing the round
+
+	// taking signature and changing the key to match different round
+	sig2 := sig
+	key, err := signer.keyStore.GetKey(start)
+	a.NoError(err)
+
+	sig2.VerifyingKey = *(key.GetSigner().GetVerifyingKey())
+	a.Error(signer.GetVerifier().Verify(start, start+1, 1, hashable, sig2))
+}
+
+func TestMarshal(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	a := require.New(t)
+
+	signer := generateTestSigner(crypto.FalconType, 0, 10, 1, a)
+	store := signer.keyStore.store
+	defer store.Close()
+
+	out := protocol.Encode(signer)
+	decodeInto := &Signer{}
+	a.NoError(protocol.Decode(out, decodeInto))
+	decodeInto.keyStore.store = store // restore PersistentKeystore
+	a.Equal(signer, decodeInto)
+
+	verifier := signer.GetVerifier()
+	bs := protocol.Encode(verifier)
+	verifierToDecodeInto := Verifier{}
+	protocol.Decode(bs, &verifierToDecodeInto)
+	a.Equal(*verifier, verifierToDecodeInto)
+}
+
+func TestSignerTrim(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	a := require.New(t)
+	var err error
+
+	signer := generateTestSigner(crypto.FalconType, 1, 100, 1, a)
+	defer signer.keyStore.store.Close()
+
+	_, err = signer.Trim(1)
+	a.NoError(err)
+	a.Equal(signer.FirstValid, uint64(1))
+	a.Equal(length(signer, a), 99)
+
+	_, err = signer.Trim(10)
+	a.NoError(err)
+	a.Equal(length(signer, a), 90)
+
+	signer.Trim(20)
+	a.Equal(length(signer, a), 80)
+
+	signer = generateTestSigner(crypto.FalconType, 1, 100, 11, a)
+	defer signer.keyStore.store.Close()
+	a.Equal(9, length(signer, a))
+
+	// Should not trim, removes only keys <= round 10
+	signer.Trim(10)
+	a.Equal(signer.FirstValid, uint64(1))
+	a.Equal(9, length(signer, a))
+
+	// Should delete keys for rounds 11 and 22
+	signer.Trim(22)
+	a.Equal(signer.FirstValid, uint64(1))
+	a.Equal(7, length(signer, a))
+
+	signer.Trim(99)
+	a.Equal(signer.FirstValid, uint64(1))
+	a.Equal(length(signer, a), 0)
+
+	// create signer and delete all keys.
+	signer = generateTestSigner(crypto.FalconType, 1, 60, 1, a)
+	defer signer.keyStore.store.Close()
+	_, err = signer.Trim(60)
+	a.NoError(err)
+	a.Equal(0, length(signer, a))
+	_, err = signer.Trim(61) // should not return error for rounds bigger than lastValid
+	a.NoError(err)
+	a.Equal(0, length(signer, a))
+
+	signer = generateTestSigner(crypto.FalconType, 1, 60, 11, a)
+	defer signer.keyStore.store.Close()
+	_, err = signer.Trim(55)
+	a.NoError(err)
+	a.Equal(0, length(signer, a))
+
```

Comment:
`Equal` first argument is the expected value, second is the checked value. The parameter order must be reversed.